File: iOS/Classes/TiGaModule.m
/**
 * Ti.GA - Google Analytics for Titanium
 * Copyright (c) 2013 by Benjamin Bahrenburg. All Rights Reserved.
 * Licensed under the terms of the MIT License
 * Please see the LICENSE included with this distribution for details.
 *
 * Available at https://github.com/benbahrenburg/Ti.GA
 *
 */
#import "TiGaModule.h"
#import "TiBase.h"
#import "TiHost.h"
#import "TiUtils.h"
#import "GAI.h"
#import "GAIFields.h"
#import "TiGaTrackerObjectProxy.h"
#import "BXBUtil.h"

@implementation TiGaModule

static BOOL _LOG_DEBUG = NO;

#pragma mark Internal

// this is generated for your module, please do not change it
-(id)moduleGUID
{
	return @"957d2892-5405-4a7f-8d44-8afaffeb17b1";
}

// this is generated for your module, please do not change it
-(NSString*)moduleId
{
	return @"ti.ga";
}

#pragma mark Lifecycle

-(void)startup
{
    _errorHandlerEnabled = NO;
	// this method is called when the module is first loaded
	// you *must* call the superclass
	[super startup];
}

-(void)shutdown:(id)sender
{	
	// you *must* call the superclass
	[super shutdown:sender];
    
    if(![NSThread isMainThread]){
        TiThreadPerformOnMainThread(^{
            [[GAI sharedInstance] dispatch];
        }, NO);
    }
}

#pragma mark Cleanup 


#pragma mark Internal Memory Management

-(void)didReceiveMemoryWarning:(NSNotification*)notification
{
	[super didReceiveMemoryWarning:notification];
}

+(BOOL) LOG_DEBUG
{
    return _LOG_DEBUG;
}

-(id)createTracker:(id)args
{
    if(args == nil){
        return [[TiGaTrackerObjectProxy alloc] initWithDefaultTracker];
    }
    
    ENSURE_SINGLE_ARG(args,NSDictionary);
    ENSURE_TYPE(args,NSDictionary);
    return [[TiGaTrackerObjectProxy alloc] initWithParams:args];
}

-(id)debug
{
    return NUMBOOL([[GAI sharedInstance].logger logLevel] == kGAILogLevelNone);
}

-(void) setDebug:(id)value
{
    if([TiUtils boolValue:value]){
        [[GAI sharedInstance].logger setLogLevel:kGAILogLevelVerbose];
        _LOG_DEBUG = YES;
        [BXBUtil logDebug:@"Debug enabled"];
    }else{
        [[GAI sharedInstance].logger setLogLevel:kGAILogLevelNone];
        _LOG_DEBUG = NO;
        [BXBUtil logDebug:@"Debug disabled, only warnings will be displayed"];
    }
}

-(id)optOut
{
    return NUMBOOL([[GAI sharedInstance] optOut]);
}
-(void)setOptOut:(id)value
{
    [[GAI sharedInstance] setOptOut:[TiUtils boolValue:value]];
}

-(id)dispatchInterval
{
    return [NSNumber numberWithDouble:[GAI sharedInstance].dispatchInterval];
}
-(void)setDispatchInterval:(id)value
{
    ENSURE_SINGLE_ARG(value,NSNumber);
    
    [BXBUtil logDebug:[@"Display level set " stringByAppendingString:[NSString stringWithFormat:@"%f",
                                                                      [TiUtils doubleValue:value]]]];
    [GAI sharedInstance].dispatchInterval = [TiUtils doubleValue:value];
}

-(NSNumber*) isTrackUncaughtExceptionsActive:(id)unused
{
    return NUMBOOL(_errorHandlerEnabled);
}
-(void) enableTrackUncaughtExceptions:(id)unused
{
    [GAI sharedInstance].trackUncaughtExceptions = YES;
    _errorHandlerEnabled = YES;
    [BXBUtil logDebug:@"trackUncaughtExceptions enabled "];
}

-(void)dispatch:(id)unused
{
    ENSURE_UI_THREAD(dispatch,unused);    
    [[GAI sharedInstance] dispatch];
    [BXBUtil logDebug:@"manual dispatch called"];
}

MAKE_SYSTEM_PROP(ANONYMIZE_IP,kGAIAnonymizeIp);
MAKE_SYSTEM_PROP(HIT_TYPE,kGAIHitType);
MAKE_SYSTEM_PROP(SESSION_CONTROL,kGAISessionControl);
MAKE_SYSTEM_PROP(NON_INTERACTION,kGAINonInteraction);
MAKE_SYSTEM_PROP(DESCRIPTION,kGAIDescription);  // synonym for kGAIScreenName
MAKE_SYSTEM_PROP(SCREEN_NAME,kGAIScreenName);   // synonym for kGAIDescription
MAKE_SYSTEM_PROP(LOCATION,kGAILocation);
MAKE_SYSTEM_PROP(REFERRER,kGAIReferrer);
MAKE_SYSTEM_PROP(PAGE,kGAIPage);
MAKE_SYSTEM_PROP(HOSTNAME,kGAIHostname);
MAKE_SYSTEM_PROP(TITLE,kGAITitle);
MAKE_SYSTEM_PROP(LANGUAGE,kGAILanguage);
MAKE_SYSTEM_PROP(TRACKING_ID,kGAITrackingId);
MAKE_SYSTEM_PROP(ENCODING,kGAIEncoding);
MAKE_SYSTEM_PROP(SCREEN_COLORS,kGAIScreenColors);
MAKE_SYSTEM_PROP(SCREEN_RESOLUTION,kGAIScreenResolution);
MAKE_SYSTEM_PROP(VIEWPORT_SIZE,kGAIViewportSize);
MAKE_SYSTEM_PROP(CLIENT_ID,kGAIClientId);

MAKE_SYSTEM_PROP(CAMPAIGN_NAME,kGAICampaignName);
MAKE_SYSTEM_PROP(CAMPAIGN_SOURCE,kGAICampaignSource);
MAKE_SYSTEM_PROP(CAMPAIGN_MEDIUM,kGAICampaignMedium);
MAKE_SYSTEM_PROP(CAMPAIGN_KEYWORD,kGAICampaignKeyword);
MAKE_SYSTEM_PROP(CAMPAIGN_CONTENT,kGAICampaignContent);
MAKE_SYSTEM_PROP(CAMPAIGN_ID,kGAICampaignId);

MAKE_SYSTEM_PROP(EVENT_CATEGORY,kGAIEventCategory);
MAKE_SYSTEM_PROP(EVENT_ACTION,kGAIEventAction);
MAKE_SYSTEM_PROP(EVENT_LABEL,kGAIEventLabel);
MAKE_SYSTEM_PROP(EVENT_VALUE,kGAIEventValue);

MAKE_SYSTEM_PROP(SOCIAL_NETWORK,kGAISocialNetwork);
MAKE_SYSTEM_PROP(SOCIAL_ACTION,kGAISocialAction);
MAKE_SYSTEM_PROP(SOCIAL_TARGET,kGAISocialTarget);


MAKE_SYSTEM_PROP(TIMING_CATEGORY,kGAITimingCategory);
MAKE_SYSTEM_PROP(TIMING_VAR,kGAITimingVar);
MAKE_SYSTEM_PROP(TIMING_VALUE,kGAITimingValue);
MAKE_SYSTEM_PROP(TIMING_LABEL,kGAITimingLabel);

MAKE_SYSTEM_PROP(APP_NAME,kGAIAppName);
MAKE_SYSTEM_PROP(APP_VERSION,kGAIAppVersion);
MAKE_SYSTEM_PROP(APP_ID,kGAIAppId);
MAKE_SYSTEM_PROP(APP_INSTALLER_ID,kGAIAppInstallerId);

MAKE_SYSTEM_PROP(EX_DESCRIPTION,kGAIExDescription);
MAKE_SYSTEM_PROP(EX_FATAL,kGAIExFatal);

MAKE_SYSTEM_PROP(CURRENCY_CODE,kGAICurrencyCode);
MAKE_SYSTEM_PROP(TRANSACTION_ID,kGAITransactionId);
MAKE_SYSTEM_PROP(TRANSACTION_AFFILIATION,kGAITransactionAffiliation);
MAKE_SYSTEM_PROP(TRANSACTION_REVENUE,kGAITransactionRevenue);
MAKE_SYSTEM_PROP(TRANSACTION_SHIPPING,kGAITransactionShipping);
MAKE_SYSTEM_PROP(TRANSACTION_TAX,kGAITransactionTax);

MAKE_SYSTEM_PROP(ITEM_PRICE,kGAIItemPrice);
MAKE_SYSTEM_PROP(ITEM_QUANTITY,kGAIItemQuantity);
MAKE_SYSTEM_PROP(ITEM_SKU,kGAIItemSku);
MAKE_SYSTEM_PROP(ITEM_NAME,kGAIItemName);
MAKE_SYSTEM_PROP(ITEM_CATEGORY,kGAIItemCategory);

MAKE_SYSTEM_PROP(USE_SECURE,kGAIUseSecure);
MAKE_SYSTEM_PROP(JAVA_ENABLED,kGAIJavaEnabled);
MAKE_SYSTEM_PROP(FLASH_VERSION,kGAIFlashVersion);
MAKE_SYSTEM_PROP(SAMPLE_RATE,kGAISampleRate);

// hit types
MAKE_SYSTEM_PROP(GAIAppView,kGAIAppView);
MAKE_SYSTEM_PROP(GAIEvent,kGAIEvent);
MAKE_SYSTEM_PROP(GAISocial,kGAISocial);
MAKE_SYSTEM_PROP(GAITransaction,kGAITransaction);
MAKE_SYSTEM_PROP(GAIItem,kGAIItem);
MAKE_SYSTEM_PROP(GAIException,kGAIException);
MAKE_SYSTEM_PROP(GAITiming,kGAITiming);

@end

File: Android/src/ti/ga/TigaModule.java
/**
 * Ti.GA - Google Analytics for Titanium
 * Copyright (c) 2013 by Ben Bahrenburg. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 * 
 * Available at https://github.com/benbahrenburg/Ti.GA
 */
package ti.ga;

import java.lang.Thread.UncaughtExceptionHandler;
import java.util.HashMap;

import org.appcelerator.kroll.KrollModule;
import org.appcelerator.kroll.annotations.Kroll;
import org.appcelerator.titanium.TiApplication;

import android.app.Activity;

import com.google.analytics.tracking.android.ExceptionReporter;
import com.google.analytics.tracking.android.Fields;
import com.google.analytics.tracking.android.GoogleAnalytics;
import com.google.analytics.tracking.android.GAServiceManager;
import com.google.analytics.tracking.android.Logger.LogLevel;
import com.google.analytics.tracking.android.Tracker;

@Kroll.module(name="Tiga", id="ti.ga")
public class TigaModule extends KrollModule
{

	// Standard Debugging variables
	public static boolean DEBUG = false;
	public static final String MODULE_SHORT_NAME = "ti.ga";	
	private GoogleAnalytics _ga = null;
	private boolean _errorHandler = false;
	
	public TigaModule()
	{
		super();
		_ga = GoogleAnalytics.getInstance(TiApplication.getInstance().getRootOrCurrentActivity());
	}

	@SuppressWarnings({"rawtypes"})
	@Kroll.method
	public TrackerObjectProxy createTracker(HashMap hm)
	{
		Util.LogDebug("calling createTracker");		
		if(hm.containsKey(Fields.TRACKING_ID)){			
			String trackerId =  hm.get(Fields.TRACKING_ID).toString();
			Util.LogDebug("Has trackingId " + trackerId);
			return new TrackerObjectProxy(_ga.getTracker(trackerId),trackerId,hm);			
		}else{
			return new TrackerObjectProxy(_ga.getDefaultTracker(),"",hm);
		}
	}
	// Properties
	@Kroll.getProperty
	public boolean getDebug()
	{
		return DEBUG;
	}
	
	@Kroll.setProperty
	public void setDebug(boolean value) {
		DEBUG = value;
		if(DEBUG){
			GoogleAnalytics.getInstance(TiApplication.getInstance().getRootOrCurrentActivity())
		    .getLogger()
		    .setLogLevel(LogLevel.VERBOSE);
			Util.LogDebug("Debugging Enabled");	
		}else{
			GoogleAnalytics.getInstance(TiApplication.getInstance().getRootOrCurrentActivity())
		    .getLogger()
		    .setLogLevel(LogLevel.WARNING);
			Util.LogDebug("Debugging Disabled, only warnings will show");	
		}
	}

	@SuppressWarnings("deprecation")
	@Kroll.method
	public void dispatch()
	{
		Util.LogDebug("dispatch called");
		GAServiceManager.getInstance().dispatchLocalHits();
	}
	
	@SuppressWarnings("deprecation")
	@Kroll.method
	public void setDispatchInterval(int interval)
	{
		Util.LogDebug("setDispatchInterval = " + interval);
		GAServiceManager.getInstance().setLocalDispatchPeriod(interval);
	}
	// Properties
	@Kroll.getProperty
	public boolean getOptOut()
	{
		return GoogleAnalytics.getInstance(TiApplication.getInstance().getRootOrCurrentActivity()).getAppOptOut();
	}
	
	@Kroll.setProperty
	public void setOptOut(boolean value) {
		Util.LogDebug("setOptOut =" + value);
		_ga.setAppOptOut(value);
	}
	
	@Kroll.method
	public boolean isTrackUncaughtExceptionsActive(){
		return _errorHandler;
	}
	
	@Kroll.method
	public void enableTrackUncaughtExceptions(){
		Tracker tracker = GoogleAnalytics.getInstance(TiApplication.getInstance().getRootOrCurrentActivity()).getDefaultTracker();
		if(tracker == null){
			Util.LogError("Tracker has not yet been created. At least one must be created for enableTrackUncaughtExceptions can be set");
			return;
		}
		UncaughtExceptionHandler gaErrHandler = new ExceptionReporter(
				tracker, // Tracker, may return null if not yet initialized.
			    GAServiceManager.getInstance(),                        // GAServiceManager singleton.
			    Thread.getDefaultUncaughtExceptionHandler(), null);          // Current default uncaught exception handler.

			// Make myHandler the new default uncaught exception handler.
			Thread.setDefaultUncaughtExceptionHandler(gaErrHandler);
			_errorHandler = true;
			Util.LogDebug("UncaughtExceptions enabled");	
	}
	
	@Override
    public void onDestroy(Activity activity) 
    {
		if(_ga!=null){
			_ga = null;
		}
        super.onDestroy(activity);
    }
	
	 @Kroll.constant public static final String ANONYMIZE_IP = Fields.ANONYMIZE_IP;
	 @Kroll.constant public static final String HIT_TYPE = Fields.HIT_TYPE;
	 @Kroll.constant public static final String SESSION_CONTROL = Fields.SESSION_CONTROL;
	 @Kroll.constant public static final String NON_INTERACTION = Fields.NON_INTERACTION;
	 @Kroll.constant public static final String DESCRIPTION = Fields.DESCRIPTION;
	 @Kroll.constant public static final String SCREEN_NAME = Fields.SCREEN_NAME;
	 @Kroll.constant public static final String LOCATION = Fields.LOCATION;
	 @Kroll.constant public static final String REFERRER = Fields.REFERRER;
	 @Kroll.constant public static final String PAGE = Fields.PAGE;
	 @Kroll.constant public static final String HOSTNAME = Fields.HOSTNAME;
	 @Kroll.constant public static final String TITLE = Fields.TITLE;
	 @Kroll.constant public static final String LANGUAGE = Fields.LANGUAGE;
	 @Kroll.constant public static final String TRACKING_ID = Fields.TRACKING_ID;
	 @Kroll.constant public static final String ENCODING = Fields.ENCODING;
	 @Kroll.constant public static final String SCREEN_COLORS = Fields.SCREEN_COLORS;
	 @Kroll.constant public static final String SCREEN_RESOLUTION = Fields.SCREEN_RESOLUTION;
	 @Kroll.constant public static final String VIEWPORT_SIZE = Fields.VIEWPORT_SIZE;
	 @Kroll.constant public static final String CLIENT_ID = Fields.CLIENT_ID;
	 @Kroll.constant public static final String CAMPAIGN_NAME = Fields.CAMPAIGN_NAME;
	 @Kroll.constant public static final String CAMPAIGN_SOURCE = Fields.CAMPAIGN_SOURCE;
	 @Kroll.constant public static final String CAMPAIGN_MEDIUM = Fields.CAMPAIGN_MEDIUM;
	 @Kroll.constant public static final String CAMPAIGN_KEYWORD = Fields.CAMPAIGN_KEYWORD;
	 @Kroll.constant public static final String CAMPAIGN_CONTENT = Fields.CAMPAIGN_CONTENT;
	 @Kroll.constant public static final String CAMPAIGN_ID = Fields.CAMPAIGN_ID;
	 @Kroll.constant public static final String EVENT_CATEGORY = Fields.EVENT_CATEGORY;
	 @Kroll.constant public static final String EVENT_ACTION = Fields.EVENT_ACTION;
	 @Kroll.constant public static final String EVENT_LABEL = Fields.EVENT_LABEL;
	 @Kroll.constant public static final String EVENT_VALUE = Fields.EVENT_VALUE;
	 @Kroll.constant public static final String SOCIAL_NETWORK = Fields.SOCIAL_NETWORK;
	 @Kroll.constant public static final String SOCIAL_ACTION = Fields.SOCIAL_ACTION;
	 @Kroll.constant public static final String SOCIAL_TARGET = Fields.SOCIAL_TARGET;
	 @Kroll.constant public static final String TIMING_CATEGORY = Fields.TIMING_CATEGORY;
	 @Kroll.constant public static final String TIMING_VAR = Fields.TIMING_VAR;
	 @Kroll.constant public static final String TIMING_VALUE = Fields.TIMING_VALUE;
	 @Kroll.constant public static final String TIMING_LABEL = Fields.TIMING_LABEL;
	 @Kroll.constant public static final String APP_NAME = Fields.APP_NAME;
	 @Kroll.constant public static final String APP_ID = Fields.APP_ID;
	 @Kroll.constant public static final String APP_INSTALLER_ID = Fields.APP_INSTALLER_ID;
	 @Kroll.constant public static final String APP_VERSION = Fields.APP_VERSION;
	 @Kroll.constant public static final String EX_DESCRIPTION = Fields.EX_DESCRIPTION;
	 @Kroll.constant public static final String EX_FATAL = Fields.EX_FATAL;
	 @Kroll.constant public static final String CURRENCY_CODE = Fields.CURRENCY_CODE;
	 @Kroll.constant public static final String TRANSACTION_ID = Fields.TRANSACTION_ID;
	 @Kroll.constant public static final String TRANSACTION_AFFILIATION = Fields.TRANSACTION_AFFILIATION;
	 @Kroll.constant public static final String TRANSACTION_SHIPPING = Fields.TRANSACTION_SHIPPING;
	 @Kroll.constant public static final String TRANSACTION_TAX = Fields.TRANSACTION_TAX;
	 @Kroll.constant public static final String TRANSACTION_REVENUE = Fields.TRANSACTION_REVENUE;
	 @Kroll.constant public static final String ITEM_PRICE = Fields.ITEM_PRICE;
	 @Kroll.constant public static final String ITEM_QUANTITY = Fields.ITEM_QUANTITY;
	 @Kroll.constant public static final String ITEM_SKU = Fields.ITEM_SKU;
	 @Kroll.constant public static final String ITEM_NAME = Fields.ITEM_NAME;
	 @Kroll.constant public static final String ITEM_CATEGORY = Fields.ITEM_CATEGORY;
	 @Kroll.constant public static final String SAMPLE_RATE = Fields.SAMPLE_RATE;
	 @Kroll.constant public static final String JAVA_ENABLED = Fields.JAVA_ENABLED;
	 @Kroll.constant public static final String FLASH_VERSION = Fields.FLASH_VERSION;
	 @Kroll.constant public static final String USE_SECURE = Fields.USE_SECURE;
	 
	 @Kroll.constant public static final String HELPER_CONSTANT_TRUE = "true";
	 @Kroll.constant public static final String HELPER_CONSTANT_FALSE = "false";
	 
	 //Android only properties
	 @Kroll.constant public static final String ANDROID_APP_UID = Fields.ANDROID_APP_UID;
	 
}

